### PR TITLE
Type inference for argument and return types of method references

### PR DESCRIPTION
**Pull request by jhorstmann**
Possible solution for issue #103 

Not integrated into the api yet because of PR #105 which refactored everything

### CODE REVIEW

**Comment by jhorstmann** [comment]
Some interesting discussion here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=449063

Sounds like it is intentional by javac to not emit the generic signature in some cases. The eclipse compiler now has a compiler option to support this.

**Comment by jhorstmann** [comment]
[More discussion and even a patch for OpenJDK][1] and an [answer by Brian Goetz][2]:

> There doesn't seem to be anything *wrong* with having a signature 
attribute on synthetic methods; the fact that we don't have them now is 
more an artifact of how erasure is implemented in the compiler than a 
deliberate decision to strip these methods of generic type information. 

 [1]: http://mail.openjdk.java.net/pipermail/compiler-dev/2015-January/009220.html
 [2]: http://mail.openjdk.java.net/pipermail/compiler-dev/2015-January/009268.html

File: src/main/java/org/zalando/riptide/TypeInference.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        if (function.getClass().isSynthetic()) {
+            return getFunctionInfoForMethodReference(function);
+        } else {
+            return getFunctionInfoForClass(function, ThrowingFunction.class);
+        }
+    }
+
+    static <T> FunctionInfo<T, Void> forConsumer(final ThrowingConsumer<T> function) {
+
+        if (function.getClass().isSynthetic()) {
```
Comment (jhorstmann) [inline src/main/java/org/zalando/riptide/TypeInference.java]: I need to create a synthetic class with custom `writeReplace` method to test code coverage here. I have no idea how to do that. Even with creating a class from bytecode, the flag seems to be cleared on class loading.

**Comment by whiskeysierra** [comment]
As discussed, we're going for a simpler API that doesn't need this.

**Pull request closed without merging**